### PR TITLE
Add source to metadata

### LINKED ISSUE #2033: Store original filename in metadata associated with layer instances
## 🚀 Feature
It would be cool if layers containing image data which were directly loaded from disc have an attribute/property containing meta data, especially the filename. 

## Motivation
I'm working on a [code generator](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_scriptgenerators/_ScriptGeneratorBase.py#L4) which takes a list of napari-layers and generates code for reproducing their content. Starting point in my plugin is a layer that has the attribute "filename".

## Pitch
By storing meta data, such as the original filename, in layers we can use it to document what has been done with images within napari. 

## Alternatives
~Attaching a general-purpose metadata dictionary to layers which can be manipulated by any plugin would be most flexible.~
Edit: @kevinyamauchi pointed me to the existing metadata dictionary. Thanks!

## Additional context
I think one could build this in [here](https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/components/viewer_model.py#L969). If we conclude on how to do it, I'd be happy to provide PR.

### PR DESCRIPTION
**Pull request by haesleinhuepf**
# Description
This PR makes napari store an image source in `layer.metadata['source']` after napari opened the image. (Closes #2033 ). Furthermore, it adds a code eample and a test to show how/that works.

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
I added a code example demonstrating how to access the meta data entry after opening an image.

# References

# How has this been tested?
- [ ] I added a test. Unfortunately, it throws a "Widgets leaked error" on my system. I'm not sure what that means. Let me know how I can improve this or if the test should live in a different place.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by alisterburt** [comment]
I think this is really great, definitely really useful! Is it worth having source as an optional keyword argument in the `Viewer.add_*` methods too?

I think this is important for those wanting to use napari on top of their projects which already presumably have data shuttling around under the hood which can't be directly opened by the viewer - if people agree I'm happy to add this in

**Comment by haesleinhuepf** [comment]
> Is it worth having source as an optional keyword argument in the `Viewer.add_*` methods too?

Great idea, indeed. I thought about that as well and decided against for now, because those methods do have a `metadata` dictionary parameter, where one could add a `source` entry.

File: napari/components/viewer_model.py [hunk 1]
```diff
@@ -995,6 +995,9 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            if 'source' not in new.metadata:
+                new.metadata['source'] = filename
```
Comment (tlambert03) [inline napari/components/viewer_model.py]:
```suggestion
            new.metadata.setdefault('source', filename)
```

good time to use [`dict.setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault)!

**Comment by alisterburt** [comment]
I left a PR on your PR which would 'promote' source from simply a piece of metadata to a property of all layers settable upon construction, linking here for others to see haesleinhuepf/napari#1

**Comment by haesleinhuepf** [comment]
> I left a PR on your PR which would 'promote' source from simply a piece of metadata to a property of all layers settable upon construction, linking here for others to see [haesleinhuepf#1](https://github.com/haesleinhuepf/napari/pull/1)

Also a good idea. I personally don't have a preference on that aspect. What do others think about this sub-PR? @tlambert03 @sofroniewn

File: napari/components/viewer_model.py (hunk 1 above)
Comment (haesleinhuepf) [reply napari/components/viewer_model.py]: Great idea!

**Comment by tlambert03** [comment]
commenting here to keep the discussion in one place...
I'm not sure I like https://github.com/haesleinhuepf/napari/pull/1 as is ... I don't love the `v.add_image(skimage.data.astronaut(), source=skimage.data.astronaut)` syntax, which requires a lot of duplication from the user.
The primary need for this PR as it currently stands is that it takes something that napari does _internally_ (specifically, taking path in the `add_path(path)` method and turning it into an `Image` layer), and leaves a little breadcrumb of what napari did.  https://github.com/haesleinhuepf/napari/pull/1, on the other hand, is just making a place for an end-user to store something that they could very easily have stored elsewhere ("hang on to this function for me would you?" :wink:).  While I don't mind the concept of remembering which function was used to generate the data, I think it would only be appropriate if we actually _accepted_ an image-generating function somewhere... which we don't (the user had to call the function before calling `add_image()`).  Since we _do_ accept a string path and turn it into image data, that original string should be accessible somewhere.

Whether source is a first-class citizen or not... that feels like something easier to add than it is to remove.  So I'm in favor of the cautious metadata approach in the short term

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -995,6 +995,8 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            new.metadata.setdefault('source', filename)
```
Comment (tlambert03) [inline napari/components/viewer_model.py]: ah right... your tests are failing because `_add_layer_from_data` can optionally return a list of layers.  That's an unfortunate design (it has to do with the fact that `add_image` can also return a list of images), but it needs to be addressed:
```python
if isinstance(new, list):
    for layer in new:
        layer.metadata.setdefault('source', filename)
else:
    new.metadata.setdefault('source', filename)
```

**Comment by alisterburt** [comment]
I see your points and guess it's not really necessary!

**Comment by sofroniewn** [comment]
> Whether source is a first-class citizen or not... that feels like something easier to add than it is to remove. So I'm in favor of the cautious metadata approach in the short term

I'm just wondering before we merge this if we've thought through the consequences of auto-populating the metadata dict like this - which has previously just been for users - and if unexpected things might happen if people do or do not have `source` already in their metadata dict, or if one day plugins want to provide their own metadata dicts. It does feel like it's bridging two worlds to have napari place things in here automatically, on the other hand without this there is no way to capture the current information.

An alternative I see is that the layer gets a `source` attribute, but I'm not so sure about that either as most layers won't need this, but it could be None for those. A `path` attribute makes more sense to me then as we have the `kwarg` path. "source" is a word we've never used anywhere else in the codebase.

**Comment by tlambert03** [comment]
> I'm just wondering before we merge this if we've thought through the consequences of auto-populating the metadata dict like this - which has previously just been for users - and if unexpected things might happen if people do or do not have source already in their metadata dict, or if one day plugins want to provide their own metadata dicts.

since @haesleinhuepf  used `setdefault`, it will only populate source if it doesn't already exist.  it will not overwrite user data.  and since we don't use it anywhere, it's hard to think of unexpected consequences.  It's basically there for a user if they want it (like @haesleinhuepf does), and ignorable/overwrite able if they don't.  

The `layer.source` attribute idea is potentially good too (that's what @alisterburt started to do in https://github.com/haesleinhuepf/napari/pull/1) ... but my thought, mentioned above is that adding that is a bigger commitment that is harder to change in the future.  So we'd want to step back and think longer about that

**Comment by sofroniewn** [comment]
>  it's hard to think of unexpected consequences. It's basically there for a user if they want it (like @haesleinhuepf does), and ignorable/overwrite able if they don't.

Say 6 months from now @haesleinhuepf has made a really cool plugin that autogenerates amazing things by looking for `source` in the metadata, someone else has written an awesome file IO plugin that preserves all the metadata from a file and dumps it into the metadata dict using the source kwarg but for something else that's not compatible with @haesleinhuepf plugin. A third person now wants to use both plugins together and things break :-(

Maybe that's ok or we tell that fileIO plugin developer "source" is a protected key and they should use something else, but I'd prefer to avoid this situation in the first place if possible, and as I see it as pretty likely to happen if we just start putting our own stuff in the `metadata` dict.

Also, if we do the `metadata` approach here, I wonder if are we going to add more things or if we're just making a special exception as this is information "lost" by napari? For example #1646 from @Czaki requests `source` and `file path` along with some other things like "origin", all of which seem reasonable, but I'm not sure if we should be adding them all too.

Using `layer.source` as a read only property that we set inside `viewer.open` and is None otherwise would avoid this. I have similar concerns about API expansion, and I don't love storing file path info on the layer, but I think a read only property might be a better place to start than populating the metadata dict.

We could also make `layer.source` a read-only dict, or small class/ namedtuple ;-)  with things like `filename`, `plugin_name`, that we populate and which could grow to include other stuff (like if a labels layer was created by clicking on a button) over time. This might be my favourite option as it containerizes everything nicely and guarantees we'll be meeting the needs of @haesleinhuepf and others for the long term.

I think the approach from @alisterburt in https://github.com/haesleinhuepf/napari/pull/1 is very exciting, but maybe hinting at a bigger picture vision around what it means for something to be a "computational source" that we're not quite ready to bite off yet.

Another approach would be to put `source` in `metadata` dict in our builtin plugin readers. This means that another fileIO plugin would still have complete control over the metadata dict, and the metadata would match exactly what that plugin provided, but I'm not sure if this would really meet the needs of @haesleinhuepf who wants to build something potentially on top of every fileIO plugin - but exactly the reason that makes me lean towards the `layer.source` approach.

Curious what others think here and if I'm being too pessimistic/ overly cautious!!

File: napari/components/viewer_model.py [hunk 3]
```diff
@@ -995,6 +995,12 @@ def _add_layers_with_plugins(
...
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            if isinstance(new, list):
+                for layer in new:
+                    layer.metadata.setdefault('source', filename)
+            else:
+                new.metadata.setdefault('source', filename)
             # some add_* methods return a List[Layer], others just a Layer
             # we want to always return a list
             added.extend(new if isinstance(new, list) else [new])
```
Comment (Czaki) [inline napari/components/viewer_model.py]: here is twice distinguishing between list and single element. Maybe simple:
```
if not isinstance(new, list):
	new = [new]
```
on begin?

**Comment by Czaki** [comment]
When using napari as n-dimensional viewer, then knowing the source of data is not important. 

But there was already discussion about change napari in imaging data processing tools (which was already discussed at the European developer meeting) then knowing the source of data is really important. I really do not like the idea of depending on a scientist notebook (the physical one). The best thing will be the possibility to generate a whole calculation graph (like presented here https://napari.zulipchat.com/#narrow/stream/212875-general/topic/node.20editor). But I'm not sure if this should be done in `Layer` and its subclass. Maybe some additional small class could be used for such a description. Then will be simpler to keep the dependency graph in memory.

**Comment by tlambert03** [comment]
@sofroniewn, I agree with pretty much everything in https://github.com/napari/napari/pull/2053#issuecomment-752844879 ... I just saw this as a maybe-not-permanent thing that would help @haesleinhuepf now, and could easily be changed in the future.  I also think a first-class attribute like @alisterburt proposed is a more "stable" fix, but would require some more time to get right (maybe that time is now though?).  I'm pretty sure this "saving of the path" thing has come up before in the past, but I can't find the link at the moment.

I'm not sure I'm yet convinced that source should be a parameter in the `Layer.__init__` signature, but perhaps a quick solution here is to just make a `layer.source` attribute that is set to `None`, and then @haesleinhuepf can change `layer.metadata.setdefault('source', filename)` to `layer.source = filename`?

**Comment by haesleinhuepf** [comment]
Just my 2 cents:

> Say 6 months from now @haesleinhuepf has made a really cool plugin that autogenerates amazing things by looking for `source` in the metadata, someone else has written an awesome file IO plugin that preserves all the metadata from a file and dumps it into the metadata dict using the source kwarg but for something else that's not compatible with @haesleinhuepf plugin. A third person now wants to use both plugins together and things break :-(

That appears unfortunate, but reasonable. As we haven't specified a list of mandatory metadata entries (yet), breaking plugins in this context is ok. 

Until further notice, plugin developers need to check if specific entries exist in this dictionary before using them.

Another and yet different example: `voxel-size`: It _should_ exists in metadata (maybe that's the most important metadata anyway) and if it's not there, things may break. That's ok IMHO 🙂

**Comment by Czaki** [comment]
> Another and yet different example: `voxel-size`: It _should_ exists in metadata (maybe that's the most important metadata anyway) and if it's not there, things may break. That's ok IMHO slightly_smiling_face

Why? I first time hear about the mandatory of `voxel-size` in `metadata`. Spacing specifies the voxel size. metadata could contain only units.

**Comment by haesleinhuepf** [comment]
> the mandatory of `voxel-size` in `metadata`.

Voxel size is not mandatory. I just used this example to explain what kind of metadata entries could be mandatory. Sorry if I wasn't clear.

**Comment by Czaki** [comment]
> > the mandatory of `voxel-size` in `metadata`.
> 
> Voxel size is not mandatory. I just used this example to explain what kind of metadata entries could be mandatory. Sorry if I wasn't clear.

I do not think that any metadata should be mandatory without converting metadata to a custom class instance. Every mandatory thing in such a basic class like `Layer` should be named property, not an element of dict.

**Comment by haesleinhuepf** [comment]
> I do not think that any metadata should be mandatory without converting metadata to a custom class instance.

Alright. So what's your opinion, should "source" be a property of Layer?

**Comment by Czaki** [comment]
`source` should be the property of `Layer` or of some additional class with a description of a layer.  

There are three cases of source:

* Data loaded from disc – it could be represented as a simple path (`str` or `Path` object). 
* Output of some method – then knowing method without parameters does not provide useful information for me. So it should contain all information about method parameters. But one of the argument could be a layer which could be later deleted by a user. 
* User modified layer (shapes, points) – which is a root, but also may bee need to know base on which layer it is created (for me it will be useful when saving a single layer to know which file is labeled)

I think that introducing some method of layer description serialization to really know the source of a layer.

**Comment by alisterburt** [comment]
I like the idea of some kind of source class (and agree with @tlambert03 that whatever is added now shouldn't necessarily immediately become a part of the constructor)

@Czaki for you the `source` seems to be the computational source of the specific data on the layer (with info to serialize it), I was also imagining the `source` might be useful for layer management. It could be cool if something like "show all layers which have the same `dataset_identifier` in their `source`" was possible in a standard way - I might be getting off track and confusing the purpose of `source` though...

**Comment by haesleinhuepf** [comment]
Huh. I'm not sure how the `layer properties` you refer to differentiate from `metadata`. If I printed the `metadata` dictionary and it didn't contain the source of the data [and its voxel size] it would appear incomplete to me.

> some additional class with a description of a layer.

This goes pretty much in line with "data that provides information about other data", which is the [definition of metadata](https://en.m.wikipedia.org/wiki/Metadata).

**Comment by sofroniewn** [comment]
> I'm not sure I'm yet convinced that source should be a parameter in the Layer.__init__ signature, but perhaps a quick solution here is to just make a layer.source attribute that is set to None, and then @haesleinhuepf can change layer.metadata.setdefault('source', filename) to layer.source = filename?

@tlambert03 this would probably be my top vote too. An attribute that is set to `None` by default. We can then figure out longer term solution with time and easily add a deprecation warning if that changes etc. One option would be to make it a `property` with no setter and have a private `_source` that we set inside our `_add_path_with_plugins` method, so `layer._source = filename`. This would make `layer.source` "read only" from the user perspective, which I think is fine for now too.

> Huh. I'm not sure how the layer properties you refer to differentiate from metadata. If I printed the metadata dictionary and it didn't contain the source of the data [and its voxel size] it would appear incomplete to me.

I think there might be some understandable confusion about how we've referred to `layer.metadata` vs other layer attributes or named properties. Things so fundamental that they are always needed for viewing the data, such as voxel size have been stored on the layer directly. You should be able to access that right now with `layer.scale` (though that is in someways overloaded as it could include both a voxel-size and an affine transform component so it's maybe not quite right, but curious if that meets your voxel size needs). (Similarly we have `layer.colormap`, `layer.contrast_limits` etc all as attributes directly on the layer).

Right now we have a private method `layer._get_state()` that will put all this stuff in a dictionary that might be more akin to the full "metadata" that you're imagining. `layer.metadata` has always just been free "user space" that is in addition to this other information, and so maybe isn't the best named. We'll be aiming to clean this up soon though so there is a public way to get this dictionary too, see #1636. Does this make it more clear @haesleinhuepf?

**Comment by haesleinhuepf** [comment]
Alrigt. That interestingly also brings us back to where [our discussion started](https://github.com/napari/napari/issues/2033#issue-772126928).

File: napari/_tests/test_adding_removing.py [hunk 4]
```diff
@@ -120,3 +120,15 @@ def my_custom_callback(event):
     assert len(layer.events.callbacks) == 1
     for em in layer.events.emitters.values():
         assert len(em.callbacks) == 1
+
+
+def test_open_metadata_source_stored(make_test_viewer):
```
Comment (sofroniewn) [inline napari/_tests/test_adding_removing.py]: I think we can just add an additional assert to this test here https://github.com/napari/napari/blob/v0.4.2/napari/components/_tests/test_add_layers.py#L45-L48 rather than making a whole new one. Or if we do make a new one it should at least be in that file and use one of our mocked paths rather than a web url

Test would get an additional line

```python
    assert viewer.layers[0].source == 'mock_path.tif'
```

I think

File: examples/open_image.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+from skimage import data
+import napari
+
+with napari.gui_qt():
+    # create the viewer
+    viewer = napari.Viewer()
+
+    # open locally stored image
+    layers = viewer.open(image_local_filename)
+    print("Image " + layers[0].metadata['source'] + " loaded.")
```
Comment (sofroniewn) [inline examples/open_image.py]:
```suggestion
    print("Image " + layers[0].source + " loaded.")
```

File: examples/open_image.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+    # open image from web
+    layers = viewer.open(image_web_url)
+    print("Image " + layers[0].metadata['source'] + " loaded.")
```
Comment (sofroniewn) [inline examples/open_image.py]:
```suggestion
    print("Image " + layers[0].source + " loaded.")
```

File: napari/components/viewer_model.py [hunk 7]
```diff
@@ -995,9 +995,16 @@ def _add_layers_with_plugins(
...
             new = self._add_layer_from_data(*_data)
+
             # some add_* methods return a List[Layer], others just a Layer
             # we want to always return a list
-            added.extend(new if isinstance(new, list) else [new])
+            new = new if isinstance(new, list) else [new]
+
+            # store the source of the layer
+            for layer in new:
+                layer.source = filename
```
Comment (sofroniewn) [inline napari/components/viewer_model.py]: I think we should add `layer.source = None` to the [base layer](https://github.com/napari/napari/blob/v0.4.2/napari/layers/base/base.py) with documentation in the attributes is always present.

Also curious how people would feel about making it a read only property with no setter and then using `layer._source = filename` internally here.

Inside the base class it would look like something like

```python
    @property
    def source(self) -> Optional[str]:
        """Return source of data if loaded from a file. None otherwise.
        """
        return self._source
```

**Comment by sofroniewn** [comment]
> Alrigt. That interestingly also brings us back to where our discussion started.

Lol, I think we're making progress!! I just left some additional comments :-)

**Comment by sofroniewn** [comment]
Bumped milestone to `0.4.3` - we should be able to get this in for next release!

**Comment by Czaki** [comment]
Maybe modify `io` plugin mechanism to check if `source` is set and if no then set to the path used to load data.

**Comment by d-v-b** [comment]
i think every array has some location -- an address in memory, a path on the local file system, a path on the cloud. are there any exceptions to this? if not, then i would consider using a URL-style for the source like `file://foo/bar.tiff`, `memory://<memory_address>`, `s3://foo/bar` etc

File: napari/components/viewer_model.py (hunk 7 above)
Comment (tlambert03) [reply napari/components/viewer_model.py]: I'm for this.  Read only good too for now.  More complex usage can come later as use cases arises

**Comment by Czaki** [comment]
The memory address is a bad idea. It may be easily invalidated. It also does not contain metadata which is the most important part when we think about semantic meaning.

**Comment by haesleinhuepf** [comment]
I just implemented @sofroniewn's suggestion with the private `_source` attribute and the public/read-only `source` property. I also agree with @tlambert03: Let's see how this works in practice, before we engineer it further towards practical use-cases.

**Comment by d-v-b** [comment]
>The memory address is a bad idea. It may be easily invalidated. 

Any path can be invalidated, even ones on the filesystem. 

> It also does not contain metadata which is the most important part when we think about semantic meaning.

Sure, the path to data is just one part of the set of metadata. My point is just that if napari keeps track of the path / location / origin for data loaded from disk, it should also keep track of the path / location / origin (or something sufficiently path-like) for all other sources of data.

File: napari/_tests/test_adding_removing.py (hunk 4 above)
Comment (sofroniewn) [reply napari/_tests/test_adding_removing.py]: can you address this comment too @haesleinhuepf, then I think it's ready to be approved

**Comment by sofroniewn** [comment]
> My point is just that if napari keeps track of the path / location / origin for data loaded from disk, it should also keep track of the path / location / origin (or something sufficiently path-like) for all other sources of data.

We can grow this over time depending on what people request/ need. I think it's ok to get a first version of this in without, but I like the idea too!!

**Comment by Czaki** [comment]
> Sure, the path to data is just one part of the set of metadata. My point is just that if napari keeps track of the path / location / origin for data loaded from disk, it should also keep track of the path / location / origin (or something sufficiently path-like) for all other sources of data.

If you keep only memory address it is worthless. Data could be dask where you dynamically load and dump data on the drive. Could be NumPy array, which may not be continuous in memory. And we do not talk about data type (np.uint8, np.uint16, np.int16). Memory pointer without information on how to read it is worthless.  Encoding such data in the string is asking for a problem for me.
There are also things like points or shapes which are python objects and the memory address approach is stupid for me (python does not guarantee memory location of the object, you keep only reference to it).  

Disc or network location is much stable than in memory. Also often the path to file or file itself contains data that allows for semantic understanding of data. 

Also much better is to have information that `layer_c` is calculated base on `layer_a` (which is loaded from `file_a`) and `layer_b` (loaded from `file_b`). than `layer_c` is calculate base on `pointer_a` and `pointer_b`.

**Comment by d-v-b** [comment]
so then maybe the conclusion is that for in-memory data napari should give up and put `None` for the source? That seems fine as long as consumers of the source field know what that means. (also happy to end this derail since it's not on topic for this PR)

**Comment by jni** [comment]
I'm not super-convinced about adding a new attribute to the layers to be honest... At least, it should be in experimental, or discussed in one or more group meetings, or both. I see your point @sofroniewn about conflicting metadata specs, but I also share @tlambert03's concern that fiddling with that dictionary is much less risky, from a user-space perspective, than fiddling with attributes in the base layer. (Sorry to have jumped in late after all this discussion has already happened.)

In short, if we're not ok with the dictionary approach (again, I see the issues), I'd like us to take some time to think about this and *not* include an API expansion in the next release. @haesleinhuepf I think for your immediate use case, you can use the dictionary in user-space as suggested by @sofroniewn?

**Comment by Czaki** [comment]
I vote only for parameters if it will be mandatory element. If all coda need to assume that this key may be missed I'm OK with putting it in dict.

**Comment by haesleinhuepf** [comment]
> I think for your immediate use case, you can use the dictionary in user-space as suggested by @sofroniewn?

Yes. [I am already using this](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_scriptgenerators/_JythonGenerator.py#L30). No hurry from my side.

I would also vote for keeping layer properties in a metadata dictionary if they are not necessary for core-functionality of napari.

**Comment by sofroniewn** [comment]
Ok, I've bumped this to `0.4.4` Let's discuss in a couple upcoming team meetings. Thanks all for their input and patience!

**Comment by tlambert03** [comment]
I'd like to pick this up again.  See long thread in zulip (but mostly [this post](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Working.20with.20image.20channels/near/233219505)).  Basically, the unfortunate situation at the moment is that, when someone (like a plugin) wants to evaluate a layer in the viewer, they basically have _no_ idea about the layer provenance (they don't know the file name or extension, they don't know which plugin loaded it, etc... ).  Given that we _also_ don't have a metadata model in napari, that pretty much leaves them in the dark.

From the looks of it, the reason this effort stalled is @jni's https://github.com/napari/napari/pull/2053#issuecomment-753242516 about the API.   So let's pick it up again and discuss the API.  I would propose that we add one new public read-only attribute to `Layer` called `layer.source`, or `layer.provenance` ...  It should minimally have the path (if relevant), but preferably be expandable, so perhaps a frozen dataclass?

```python
@dataclass(frozen=True)
class Source:
    path: Optional[str]  # if `viewer.open` was used
    method: Optional[MethodType]  # could be a pointer to `viewer.open`, or `viewer.add_image`
    hookimpl: Optional[HookImplementation]  # all the necessary plugin info
    plugin_name: Optional[str]  # contained in hookimpl, but could add here too?
```

and then napari can internally populate `layer._source` depending on the load method.  I'm happy to push this through once we agree on an API.  @jni, @sofroniewn

**Comment by sofroniewn** [comment]
I'm happy to pick up this thread again, that exchange in zulip was a good motivator.

I like the public readonly `layer.source` with `source` also a `kwarg` to the layer init. It would also be something that a reader plugin could provide. One thing to note with readonly is that we might run into this problem if we don't have a setter and a magicgui widget wants to update the source in place - see https://github.com/napari/napari/pull/2507#issuecomment-813676702. So maybe having a setter is ok and we eventually use our concept of locked/ frozen #1570 to prevent editing.

>  It should minimally have the path (if relevant), but preferably be expandable, so perhaps a frozen dataclass?

I like this! With those attributes as well. My comment from above might apply too though about allowing it to be editable in the case of a layer that updates in place. Maybe then the whole source object needs to be set new though?

Overall though I like this general direction and think it will add a lot of value

**Comment by tlambert03** [comment]
> I like the public readonly layer.source with source also a kwarg to the layer init. It would also be something that a reader plugin could provide.

Actually... I was thinking this should start as something that _doesn't_ have an `__init__` kwarg, and _couldn't_ be modified by plugins. part of the whole problem here started with the "unknown" that is plugins.  So this is something that we can, for now at least, strictly regulate to say something about the provenance of data.

and yeah, if `layer._source` needed to be changed, we internally would create a new `Source()` object and replace the old one

**Comment by sofroniewn** [comment]
> Actually... I was thinking this should start as something that doesn't have an __init__ kwarg, and couldn't be modified by plugins. part of the whole problem here started with the "unknown" that is plugins. So this is something that we can, for now at least, strictly regulate to say something about the provenance of data.

lol! ok nice, I'm remembering this now - we would auto-populate this stuff based on where the data came from, like we currently do in this PR with `layer._source = filename`.

We could expand that to have those other attributes, but we as napari would set that, not a plugin/ or user. (Another one I thought of is a `drag_and_drop` `bool`, that we'd set to be `True` if the data came in via drag and drop).

I like this approach, much better than having kwargs/ or setter.

I guess from my perspective then the only remaining question is should `source` just be that path or the frozen dataclass, and if dataclass what are the attributes? I like the dataclass idea, and like the ones you listed!!

**Comment by sofroniewn** [comment]
This has been superseded by #2518. Excited to see what comes next!!
